File: ALIPHATIC'S ASSIGNMENT/EXERCISE-1/PYTHON/HEAP.py
# EXERCISE 1 ON HEAP

# Implement a heap data structure storing numeric data in your chosen language with the following operations
# heapify() - balance a heap 
# insert(item) - add a new item to a heap
# find(item) - returns the node containing search item.
# findParent(item) - returns parent of an item
# findChildren(item) - returns children of an item
# remove(item) - remove an item
# traverse() - print out values from top to down
# heapsort() - sort the items of a heap



class Heap:
    def __init__(self):
        self.heap = []

    def heapify(self, arr):
        self.heap = arr
        for i in range(len(self.heap)//2, -1, -1):
            self.__min_heapify(i)

    def insert(self, item):
        self.heap.append(item)
        self.__percolate_up(len(self.heap)-1)

    def find(self, item):
        for i in range(len(self.heap)):
            if self.heap[i] == item:
                return i
        return -1

    def findParent(self, item):
        for i in range(len(self.heap)):
            if self.heap[i] == item:
                return self.heap[(i-1)//2]
        return -1

    def findChildren(self, item):
        for i in range(len(self.heap)):
            if self.heap[i] == item:
                return self.heap[2*i+1], self.heap[2*i+2]
        return -1

    def remove(self, item):
        for i in range(len(self.heap)):
            if self.heap[i] == item:
                self.heap[i], self.heap[-1] = self.heap[-1], self.heap[i]
                self.heap.pop()
                self.__min_heapify(i)
                return
        print("Item not found")

    def traverse(self):
        for i in self.heap:
            print(i, end=" ")
        print()

    def heapsort(self):
        for i in range(len(self.heap)-1, -1, -1):
            self.heap[0], self.heap[i] = self.heap[i], self.heap[0]
            self.__min_heapify(0, i-1)

    def __percolate_up(self, index):
        parent = (index-1)//2
        if index <= 0:
            return
        elif self.heap[parent] > self.heap[index]:
            self.heap[parent], self.heap[index] = self.heap[index], self.heap[parent]
            self.__percolate_up(parent)

    def __min_heapify(self, index, size=None):
        if size is None:
            size = len(self.heap)-1
        left = 2*index+1
        right = 2*index+2
        smallest = index
        if left <= size and self.heap[left] < self.heap[index]:
            smallest = left
        if right <= size and self.heap[right] < self.heap[smallest]:
                smallest = right
        if smallest != index:
            self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
            self.__min_heapify(smallest, size)


if __name__ == "__main__":
    heap = Heap()
    heap.heapify([1, 2, 3, 4, 5, 6, 7, 8, 9])
    heap.insert(10)
    heap.insert(11)
    heap.insert(12)
    heap.insert(13)
    heap.insert(14)
    heap.insert(15)
    heap.insert(16)
    heap.insert(17)
    heap.insert(18)
    heap.insert
    heap.traverse()
    heap.remove(10)
    heap.traverse()
    heap.heapsort()
    heap.traverse()
    print(heap.find(10))
    print(heap.findParent(10))
    print(heap.findChildren(10))



        
                                   
                                

File: Tunde_Wey/heap.js
class MinHeap {
  constructor() {
    this.heap = [];
  }

  heapify() {
    const n = this.heap.length;
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
      this.heapifyDown(i);
    }
  }

  heapifyDown(index) {
    const n = this.heap.length;
    let smallest = index;
    const left = 2 * index + 1;
    const right = 2 * index + 2;

    if (left < n && this.heap[left] < this.heap[smallest]) {
      smallest = left;
    }

    if (right < n && this.heap[right] < this.heap[smallest]) {
      smallest = right;
    }

    if (smallest !== index) {
      this.swap(index, smallest);
      this.heapifyDown(smallest);
    }
  }

  insert(item) {
    this.heap.push(item);
    this.heapifyUp(this.heap.length - 1);
  }

  heapifyUp(index) {
    const parent = Math.floor((index - 1) / 2);
    if (index > 0 && this.heap[index] < this.heap[parent]) {
      this.swap(index, parent);
      this.heapifyUp(parent);
    }
  }

  find(item) {
    const index = this.heap.indexOf(item);
    return index !== -1 ? index : null;
  }

  findParent(item) {
    const index = this.find(item);
    if (index !== null && index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      return this.heap[parentIndex];
    }
    return null;
  }

  findChildren(item) {
    const index = this.find(item);
    if (index !== null) {
      const leftIndex = 2 * index + 1;
      const rightIndex = 2 * index + 2;
      const leftChild =
        this.heap[leftIndex] !== undefined ? this.heap[leftIndex] : null;
      const rightChild =
        this.heap[rightIndex] !== undefined ? this.heap[rightIndex] : null;
      return [leftChild, rightChild];
    }
    return null;
  }

  remove(item) {
    const index = this.find(item);
    if (index !== null) {
      this.swap(index, this.heap.length - 1);
      this.heap.pop();
      this.heapifyDown(index);
    }
  }

  swap(i, j) {
    const temp = this.heap[i];
    this.heap[i] = this.heap[j];
    this.heap[j] = temp;
  }

  traverse() {
    return this.heap;
  }

  heapsort() {
    const sorted = [];
    const heapCopy = [...this.heap];

    while (heapCopy.length) {
      sorted.push(heapCopy[0]);
      this.swap(0, heapCopy.length - 1);
      heapCopy.pop();
      this.heapifyDown(0);
    }

    return sorted;
  }
}

// Example usage:
const heap = new MinHeap();
heap.insert(5);
heap.insert(3);
heap.insert(8);
heap.insert(2);
heap.insert(1);
console.log(heap.traverse()); // Output: [1, 2, 8, 5, 3]
console.log(heap.find(8)); // Output: 2
console.log(heap.findParent(8)); // Output: 1
console.log(heap.findChildren(8)); // Output: [5, 3]
heap.remove(2);
console.log(heap.traverse()); // Output: [1, 3, 8, 5]
console.log(heap.heapsort()); // Output: [1, 3, 5, 8]
